webapp/components/Map/KmlUploader.js

```diff
@@ -0,0 +1,142 @@
+import './KmlUploader.scss'
+
+import React, { useEffect, useState } from 'react'
+import { useMap } from 'react-leaflet'
+
+import * as User from '@core/user/user'
+import { useUser } from '@webapp/store/user'
+import { useI18n } from '@webapp/store/system'
+
+import L from 'leaflet'
+//import * as KML from './L.KML'
+require('./L.KML')
+
+import JSZip from 'jszip'
+import shp from 'shpjs'
+
+export const KmlUploader = () => {
+  const map = useMap()
+  const jszip = new JSZip()
+
+  const i18n = useI18n()
+  const user = useUser()
+
+  const [selectedFile, setSelectedFile] = useState()
+  const [isFilePicked, setIsFilePicked] = useState(false)
+  const [layers, setLayers] = useState([])
+  const [opacity, setOpacity] = useState(50)
+
+  useEffect(() => {
+    if (isFilePicked) {
+      if (selectedFile.name.endsWith('.kmz')) {
+        handleKMZ(selectedFile)
+      } else if (selectedFile.name.endsWith('.kml')) {
+        const reader = new FileReader()
+        reader.onload = async (e) => {
+          const text = e.target.result
+          addKMLLayers([text])
+        }
+        reader.readAsText(selectedFile)
+      } else if (selectedFile.name.endsWith('.zip')) {
+        const geo = L.geoJson(
+          { features: [] },
+          {
+            onEachFeature: function popUp(f, l) {
+              var out = []
+              if (f.properties) {
+                for (var key in f.properties) {
+                  out.push(key + ': ' + f.properties[key])
+                }
```

Comment:
it could be more readable with something like this: 
const out = Object.entries(f.properties).map(...

Reply:
Oh yeah, been doing so much c# that didn't automatically think in a functional way, which I should in modern js